### DISCUSSION TITLE
Dynamically set `jobs.<job_id>.needs`

### DISCUSSION THREAD

**Discussion by victorlin** [community/community#38492] [Actions]
Currently, `jobs.<job_id>.needs` can only be set as a literal array. I would like to set it dynamically so the job dependencies within the workflow can adapt to different scenarios determined by variables.

I have tried the following:

1. Conditionally add an entry to the array.

    ```yaml
    needs:
    - job1
    - ${{ true && '' || 'job2' }}
    ```

    [Error](https://github.com/victorlin/github-test/actions/runs/3412667448/workflow):

    ```
    A template expression is not allowed in this context
    ```

2. Dynamically set the array for `needs`.

    ```yaml
    needs: ${{ true && [job1] || [job1, job2] }}
    ```

    [Error](https://github.com/victorlin/github-test/actions/runs/3412719267/workflow):

    ```
    Unexpected symbol: '['. Located at position 9 within expression: true && [job1] || [job1, job2]
    ```

3. Replace `true` with a value.

    ```yaml
    needs: ${{ github.event.inputs.run_parallel && [job1] || [job1, job2] }}
    ```

    [Error](https://github.com/victorlin/github-test/actions/runs/3412746505/workflow):

    ```
    Unrecognized named-value: 'github'. Located at position 1 within expression: github.event.inputs.run_parallel && [job1] || [job1, job2]
    ```

**Comment by ilmax** [comment]
I'm also facing the problem, is there any known workaround here?

**Reply by victorlin** [reply]
I use two separate jobs with mutually exclusive job-level `if` conditions, so that one is always skipped per workflow run. This results in a bit of duplication, but you could minimize the amount of duplication by calling out to another re-usable workflow/action.

**Comment by csm10495** [comment]
I have a similar case where I'd like to have a 'golden flow' that calls a set of other actions, but also allow certain steps to be conditionally removed.

So normally
```
A -> B -> C
```

but if someone opts out of step B, it would go

```
A -> C
```

not really sure how to do that without being able to dynamically set `needs`

**Reply by karolba** [reply]
You can do that by:

- skipping the `B` step conditionally,
- keeping `needs: [A, B]` in C
- adding `if: ${{ !(failure() || cancelled()) }}` to C (and all jobs that `need` C), so that if B is skipped the rest of the workflow still runs

**Reply by nazarEnzo** [reply]
That really solved the issue. Btw there is no need to add this condition to all jobs dependant on `C` if `C` is always executed.

**Reply by sh977218** [reply]
this is not true,  if C is depend on B, and B is skipped, not having `if: ${{ !(failure() || cancelled()) }}` on C will skip C.

**Reply by nazarEnzo** [reply]
You are correct, but what I said is that you don't need to add this condition to all the jobs that depend on C (since C is always executed)

> this is not true, if C is depend on B, and B is skipped, not having `if: ${{ !(failure() || cancelled()) }}` on C will skip C.

**Reply by sh977218** [reply]
I tried having only `if: ${{ !(failure() || cancelled()) }}` on C, and the job D, which depends on C, does not run, shows as skipped.

**Reply by nazarEnzo** [reply]
I am not sure why it is not like this for you, but i specifically tested this scenario and only wrote my comment after confirming it - that is also how my current CI workflow works. Maybe job d also depends on the previous ones or job C is also optional?

**Only the jobs which directly depend on the optional job need to have this condition:
`if: ${{ !(failure() || cancelled()) }}`**
